File: Data Structures/#07.Tree Data Structures Exercises/Java/src/main/java/BinarySearchTree.java
import java.util.*;
import java.util.function.Consumer;

public class BinarySearchTree<T extends Comparable<T>> {

    private Node root;
    private int nodesCount;

    public BinarySearchTree() {
    }

    private BinarySearchTree(Node root) {
        this.preOrderCopy(root);
    }

    private void preOrderCopy(Node node) {
        if (node == null) {
            return;
        }

        this.insert(node.value);
        this.preOrderCopy(node.left);
        this.preOrderCopy(node.right);
    }

    public Node getRoot() {
        return this.root;
    }

    public int getNodesCount() {
        return this.nodesCount;
    }

    public void insert(T value) {
        this.nodesCount++;

        if (this.root == null) {
            this.root = new Node(value);
            return;
        }

        Node parent = null;
        Node current = this.root;
        while (current != null) {
            parent = current;
            parent.childrenCount++;

            if (value.compareTo(current.value) < 0) {
                current = current.left;
            } else if (value.compareTo(current.value) > 0) {
                current = current.right;
            } else {
                return;
            }
        }

        Node newNode = new Node(value);
        if (value.compareTo(parent.value) < 0) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }
    }

    public boolean contains(T value) {
        Node current = this.root;
        while (current != null) {
            if (value.compareTo(current.value) < 0) {
                current = current.left;
            } else if (value.compareTo(current.value) > 0) {
                current = current.right;
            } else {
                break;
            }
        }

        return current != null;
    }

    public BinarySearchTree<T> search(T item) {
        Node current = this.root;
        while (current != null) {
            if (item.compareTo(current.value) < 0) {
                current = current.left;
            } else if (item.compareTo(current.value) > 0) {
                current = current.right;
            } else {
                break;
            }
        }

        return new BinarySearchTree<>(current);
    }

    public void eachInOrder(Consumer<T> consumer) {
        this.eachInOrder(this.root, consumer);
    }

    private void eachInOrder(Node node, Consumer<T> consumer) {
        if (node == null) {
            return;
        }

        this.eachInOrder(node.left, consumer);
        consumer.accept(node.value);
        this.eachInOrder(node.right, consumer);
    }

    public Iterable<T> range(T from, T to) {
        Deque<T> queue = new LinkedList<>();
        this.range(this.root, queue, from, to);
        return queue;
    }

    private void range(Node node, Deque<T> queue, T startRange, T endRange) {
        if (node == null) {
            return;
        }

        int compareStart = startRange.compareTo(node.value);
        int compareEnd = endRange.compareTo(node.value);
        if (compareStart < 0) {
            this.range(node.left, queue, startRange, endRange);
        }
        if (compareStart <= 0 && compareEnd >= 0) {
            queue.addLast(node.value);
        }
        if (compareEnd > 0) {
            this.range(node.right, queue, startRange, endRange);
        }
    }

    private T minValue(Node root) {
        T minv = root.value;
        while (root.left != null) {
            minv = root.left.value;
            root = root.left;
        }

        return minv;
    }

    public void deleteMin() {
        if (this.root == null) {
            throw new IllegalArgumentException("Tree is empty!");
        }

        Node min = this.root;
        Node parent = null;

        while (min.left != null) {
            parent = min;
            parent.childrenCount--;
            min = min.left;
        }

        if (parent == null) {
            this.root = this.root.right;
        } else {
            parent.left = min.right;
        }

        this.nodesCount--;
    }

    public void deleteMax() {
        if (this.root == null) {
            throw new IllegalArgumentException("Tree is empty!");
        }

        if (this.root.right == null) {

            if (this.root.left != null) {
                this.root = this.root.left;
                return;
            }

            this.root = null;
            return;

        }

        Node fatherElement = this.root;
        Node rightElement = this.root.right;

        while (rightElement != null) {

            if (rightElement.right == null) {
                fatherElement.right = rightElement.left;
                break;
            }

            fatherElement = rightElement;
            rightElement = rightElement.right;

        }
        this.nodesCount--;
    }

    public T ceil(T element) {

        if(this.root == null){
            return null;
        }

        Queue<Node> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        Node currentHighest = null;
        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.value.compareTo(element) >= 0) {

                if (currentHighest == null) {
                    currentHighest = node;
                } else if (currentHighest.value.compareTo(node.value) >= 0) {
                    currentHighest = node;
                }
            }

            if (node.left != null)
                nodes.add(node.left);

            if (node.right != null)
                nodes.add(node.right);

            nodes.poll();
        }

        if (currentHighest == null)
            return null;

        return currentHighest.value;
    }

    //BFS
    public T floor(T element) {
        Queue<Node> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        Node currentSmallest = null;
        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.value.compareTo(element) <= 0) {

                if (currentSmallest == null) {
                    currentSmallest = node;
                } else if (currentSmallest.value.compareTo(node.value) <= 0) {
                    currentSmallest = node;
                }
            }

            if (node.left != null)
                nodes.add(node.left);

            if (node.right != null)
                nodes.add(node.right);

            nodes.poll();
        }

        if (currentSmallest == null)
            return null;

        return currentSmallest.value;
    }

    public void delete(T key) {

        if (this.search(key).root == null) {
            return;
        }

        if (this.root.childrenCount == 1) {
            this.root = null;
            return;
        }

        if (this.root.value.equals(key)) {

            if (this.root.left == null) {
                this.root = this.root.right;
                return;
            }

            if (this.root.right == null) {
                this.root = this.root.left;
                return;
            }

            Node newRoot = this.root.right;

            Node edge = newRoot;
            while (edge != null){
                Node nextEdge = edge.left;

                if (nextEdge == null) {
                    edge.left = this.root.left;
                    break;
                }

                edge = edge.left;
            }

            this.root = newRoot;
            return;
        }

        this.delete(this.root, null, key, null);
    }

    private void delete(Node startNode, Node father, T element, String comingPosition) {

        while (startNode != null) {

            if (startNode.value.equals(element)) {
                switch (comingPosition) {
                    case "left":

                        father.left = startNode.right;

                        Node lastElement = father.left;

                        if (lastElement == null)
                            father.left = startNode.left;

                        while (lastElement != null) {
                            Node next = lastElement.left;
                            if (next == null) {
                                lastElement.left = startNode.left;
                                break;
                            }
                            lastElement = lastElement.left;
                        }
                        break;

                    case "right":
                        father.right = startNode.left;

                        Node lastRightElement = father.right;

                        if (lastRightElement == null)
                            father.right = startNode.right;

                        while (lastRightElement != null) {
                            Node next = lastRightElement.right;
                            if (next == null) {
                                lastRightElement.right = startNode.right;
                                break;
                            }
                            lastRightElement = lastRightElement.right;
                        }
                        break;
                }
            }

            this.delete(startNode.left, startNode, element, "left");
            this.delete(startNode.right, startNode, element, "right");
            return;
        }
    }

    public int rank(T item) {

        List<Node> nodes = new ArrayList<>();
        this.rank(this.root, item, nodes);

        return nodes.size();
    }

    private void rank(Node node, T smallerThan, List<Node> nodes) {

        while (node != null) {

            if (node.value.compareTo(smallerThan) < 0) {
                nodes.add(node);
            }

            if (node.value.compareTo(smallerThan) <= 0) {
                this.rank(node.left, smallerThan, nodes);
                this.rank(node.right, smallerThan, nodes);
            } else {
                this.rank(node.left, smallerThan, nodes);
            }

            break;
        }
    }

    public T select(int n) {
        Map<T, Integer> map = new LinkedHashMap<>();
        this.select(this.root, map);

        return map.entrySet().stream().filter(x -> x.getValue() == n).findFirst().orElse(null).getKey();
    }

    private void select(Node node, Map<T, Integer> map) {

        while (node != null) {
            this.select(node.left, map);

            List<Node> nodes = new ArrayList<>();
            this.rank(this.root, node.value, nodes);

            map.put(node.value, (nodes.size()));
            //System.out.println(node.value + " -> " + (nodes.size() - 1));

            this.select(node.right, map);
            return;
        }
    }

    class Node {
        private T value;
        private Node left;
        private Node right;

        private int childrenCount;

        public Node(T value) {
            this.value = value;
            this.childrenCount = 1;
        }

        public T getValue() {
            return this.value;
        }

        public void setValue(T value) {
            this.value = value;
        }

        public Node getLeft() {
            return this.left;
        }

        public void setLeft(Node left) {
            this.left = left;
        }

        public Node getRight() {
            return this.right;
        }

        public void setRight(Node right) {
            this.right = right;
        }

        public int getChildrenCount() {
            return childrenCount;
        }

        public void setChildrenCount(int childrenCount) {
            this.childrenCount = childrenCount;
        }

        @Override
        public String toString() {
            return this.value + "";
        }
    }
}

File: Data Structures/#07.Tree Data Structures Exercises/C#/BinarySearchTree/BinarySearchTree.cs
using System;
using System.Collections.Generic;

public class BinarySearchTree<T> : IBinarySearchTree<T> where T : IComparable
{
    private Node root;

    private Node FindElement(T element)
    {
        Node current = this.root;

        while (current != null)
        {
            if (current.Value.CompareTo(element) > 0)
            {
                current = current.Left;
            }
            else if (current.Value.CompareTo(element) < 0)
            {
                current = current.Right;
            }
            else
            {
                break;
            }
        }

        return current;
    }

    private void PreOrderCopy(Node node)
    {
        if (node == null)
        {
            return;
        }

        this.Insert(node.Value);
        this.PreOrderCopy(node.Left);
        this.PreOrderCopy(node.Right);
    }

    private Node Insert(T element, Node node)
    {
        if (node == null)
        {
            node = new Node(element);
        }
        else if (element.CompareTo(node.Value) < 0)
        {
            node.Left = this.Insert(element, node.Left);
            node.Count++;
        }
        else if (element.CompareTo(node.Value) > 0)
        {
            node.Right = this.Insert(element, node.Right);
            node.Count++;
        }

        return node;
    }

    private void Range(Node node, Queue<T> queue, T startRange, T endRange)
    {
        if (node == null)
        {
            return;
        }

        int nodeInLowerRange = startRange.CompareTo(node.Value);
        int nodeInHigherRange = endRange.CompareTo(node.Value);

        if (nodeInLowerRange < 0)
        {
            this.Range(node.Left, queue, startRange, endRange);
        }
        if (nodeInLowerRange <= 0 && nodeInHigherRange >= 0)
        {
            queue.Enqueue(node.Value);
        }
        if (nodeInHigherRange > 0)
        {
            this.Range(node.Right, queue, startRange, endRange);
        }
    }

    private void EachInOrder(Node node, Action<T> action)
    {
        if (node == null)
        {
            return;
        }

        this.EachInOrder(node.Left, action);
        action(node.Value);
        this.EachInOrder(node.Right, action);
    }

    private BinarySearchTree(Node node)
    {
        this.PreOrderCopy(node);
    }

    public BinarySearchTree()
    {
    }

    public void Insert(T element)
    {
        this.root = this.Insert(element, this.root);
    }

    public bool Contains(T element)
    {
        Node current = this.FindElement(element);

        return current != null;
    }

    public void EachInOrder(Action<T> action)
    {
        this.EachInOrder(this.root, action);
    }

    public BinarySearchTree<T> Search(T element)
    {
        Node current = this.FindElement(element);

        return new BinarySearchTree<T>(current);
    }

    public void DeleteMin()
    {
        Stack<Node> path = new Stack<Node>();
        if (this.root == null)
        {
            return;
        }

        Node current = this.root;
        Node parent = null;
        while (current.Left != null)
        {
            parent = current;
            path.Push(parent);
            current = current.Left;
        }

        if (parent == null)
        {
            this.root = this.root.Right;
        }
        else
        {
            parent.Left = current.Right;
        }

        while (path.Count != 0)
        {
            path.Pop().updateCount();
        }
    }

    public IEnumerable<T> Range(T startRange, T endRange)
    {
        Queue<T> queue = new Queue<T>();

        this.Range(this.root, queue, startRange, endRange);

        return queue;
    }

    public void Delete(T element)
    {
        if (this.root == null)
        {
            throw new InvalidOperationException();
        }

        bool isDeleted = false;
        Stack<Node> path = new Stack<Node>();
        Queue<Node> nodes = new Queue<Node>();
        nodes.Enqueue(this.root);
        path.Push(nodes.Peek());

        while (nodes.Count != 0)
        {
            Node node = nodes.Peek();

            if (node != null)
            {
                if (node.Value.Equals(element))
                {
                    isDeleted = true;
                    if (node.Right != null)
                    {
                        Node temp = node.Left;
                        node.Value = node.Right.Value;
                        node.Right = node.Right.Right;
                        node.Left = temp;
                        node.updateCount();
                        break;
                    }
                    node.Value = node.Left.Value;
                    node.Left = node.Left.Left;
                    node.updateCount();
                    break;
                }

                if (node.Value.CompareTo(element) < 0)
                {
                    nodes.Enqueue(node.Right);
                    path.Push(node.Right);
                }
                else
                {
                    nodes.Enqueue(node.Left);
                    path.Push(node.Left);
                }
            }
            nodes.Dequeue();
        }

        if (!isDeleted)
        {
            throw new InvalidOperationException();
        }

        while (path.Count != 0)
        {
            path.Pop().updateCount();
        }
    }

    public void DeleteMax()
    {
        if (this.root == null)
        {
            throw new InvalidOperationException();
        }

        if (this.root.Right == null)
        {
            this.root = this.root.Left;
            this.root.updateCount();
            return;
        }

        Stack<Node> path = new Stack<Node>();
        Queue<Node> nodes = new Queue<Node>();
        nodes.Enqueue(this.root);
        path.Push(nodes.Peek());

        while (nodes.Count != 0)
        {
            Node node = nodes.Peek();

            if (node.Right.Right == null)
            {
                node.Right = node.Right.Left;
                break;
            }

            nodes.Enqueue(node.Right);
            path.Push(node.Right);
            nodes.Dequeue();
        }

        while (path.Count != 0)
        {
            path.Pop().updateCount();
        }
    }

    public int Count()
    {
        if (this.root == null)
        {
            return 0;
        }

        return this.root.Count;
    }

    public int Rank(T element)
    {

        int count = 0;
        Queue<Node> nodes = new Queue<Node>();
        nodes.Enqueue(this.root);

        while (nodes.Count != 0)
        {

            Node node = nodes.Peek();

            if (node != null)
                if (element.CompareTo(node.Value) > 0)
                {
                    count++;
                    nodes.Enqueue(node.Left);
                    nodes.Enqueue(node.Right);
                }
                else
                {
                    nodes.Enqueue(node.Left);
                }
            nodes.Dequeue();
        }

        return count;
    }

    public T Select(int rank)
    {
        Queue<Node> nodes = new Queue<Node>();
        nodes.Enqueue(this.root);

        while (nodes.Count != 0)
        {
            Node node = nodes.Peek();

            int result = this.Rank(node.Value);

            if (result.Equals(rank))
            {
                return node.Value;
            }

            if (node.Left != null)
                nodes.Enqueue(node.Left);

            if (node.Right != null)
                nodes.Enqueue(node.Right);

            nodes.Dequeue();
        }

        return default(T);
    }

    public T Ceiling(T element)
    {
        throw new NotImplementedException();
    }

    public T Floor(T element)
    {
        throw new NotImplementedException();
    }

    private class Node
    {

        public Node(T value)
        {
            this.Value = value;
            this.Count = 0;
        }

        public T Value { get; set; }
        public Node Left { get; set; }
        public Node Right { get; set; }

        public void updateCount()
        {
            int leftCount = this.Left != null ? this.Left.Count + 1 : 0;
            int rightCount = this.Right != null ? this.Right.Count + 1 : 0;

            this.Count = leftCount + rightCount;
        }


        public int Count { get; set; }

    }
}